File: array/BinarySearch/exact_match/cpp/BinarySearchExact.cpp
#include "BinarySearchExact.hpp"

int BinarySearchExact(const std::vector<long long>& values, long long target) {
    int left = 0;
    int right = static_cast<int>(values.size()) - 1;

    while (left <= right) {
        const int mid = left + (right - left) / 2;

        if (values[mid] == target) {
            return mid;
        }

        if (values[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return -1;
}

File: array/BinarySearch/exact_match/rust/src/binary_search_exact.rs
pub fn binary_search_exact(values: &[i64], target: i64) -> i32 {
    let mut left: i32 = 0;
    let mut right: i32 = values.len() as i32 - 1;

    while left <= right {
        let mid = left + (right - left) / 2;
        let value = values[mid as usize];

        if value == target {
            return mid;
        }

        if value < target {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    -1
}
